sensors: lis2dh: disable SDO/SA0 pull-up in case of LIS3DH

mniestroj opened the pull request:
LIS3DH contains CTRL_REG0 (1Eh) register to control internal pull-up on
SDO/SA0 line (enabled by default). Add a helper
LIS2DH_HAS_SDO_SA0_PULL_UP macro, which is defined only when "st,lis3dh"
was passed to compatible list in device-tree. Based on that macro
disable internal pull-up registers, which consume around 180uA at 3.6V
in accelerometer power-down mode.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
sensors: lis2dh: compile lis2dh_reg_field_update() unconditionally

There is little reason to compile lis2dh_reg_field_update() function
conditionally, based on enabled features. If it is not used, then linker
will drop it anyway.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj committed:
sensors: lis2dh: support disabling SDO/SA0 pull-up during init

Some chips supported by lis2dh driver (such as LIS2DH12 and LIS3DH)
contain CTRL_REG0 (1Eh) register to control internal pull-up on SDO/SA0
line (enabled by default). Add disconnect-sdo-sa0-pull-up boolean
device-tree property to allow disconnecting pull-up during driver
initialization. This allows to save around 180uA at 3.6V in
accelerometer power-down mode.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

avisconti commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -64,7 +63,6 @@ int lis2dh_reg_field_update(struct device *dev, u8_t reg_addr,
 	return lis2dh_reg_write_byte(dev, reg_addr,
 				     (old_val & ~mask) | ((val << pos) & mask));
 }
-#endif
```
There will be a compilation warning though...
Is there a reason to do it? Or is it just for the sake to keep the code more logical?

mniestroj replied:
Which compilation warning excactly? It is not a static function, so there should be no such. This function will be thrown away on the link stage.

The main reason is to get rid of noisy #ifdef statements and maintaining them for all use cases. After the change we (developers) no longer need to think about every possible compilation option, but we will depend on linker to remove the function when it is really never used.

avisconti replied:
ok, got it.

avisconti commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -321,6 +321,16 @@ int lis2dh_init(struct device *dev)
 		return status;
 	}
 
+#ifdef LIS2DH_HAS_SDO_SA0_PULL_UP
```
Why not testing DT_INST_0_ST_LIS3DH directly?

avisconti replied:
Now I also understand the previous commit... :D

mniestroj replied:
We can test LIS3DH directly. However, there are also other models which have the same "feature". One of them is already supported LIS2DH12. However there are no distinct bindings for it yet, so there is only check for LIS3DH for now.

Another reason to introducing `LIS2DH_HAS_SDO_SA0_PULL_UP` macro is to read the code in a more logical way... we disable SDO/SA0 pull-up only when the accelerometer has such feature.

avisconti replied:
Seems legit. But take a look also to the possibility to have the pullup enable/disable in Kconfig (see below).

mniestroj replied:
@pabigot Thanks for review!
Basically you suggest that disconnecting should be fully controllable by user by specifying (or not) `disconnect-sdo-sa0-pull-up`. No automatic disconnect based on compatible string. I am ok with that. However in such case I do not like the idea of additional guarding based on compatible string. This would increase maintenance effort when new models would be added. Additionally I think we should trust the user when specyfying `disconnect-sdo-sa0-pull-up` property and just do what he requests us to do (disconnect pull-up).

pabigot replied:
To be explicit: I'm fine with making it:
```
if (IS_ENABLED(DT_INST0_ST_LIS2DH_DISCONNECT_SDO_SDA0_PULL_UP)) {
   // code as below.
}
```
and letting anybody who sets that property on a device that doesn't support it deal with any side effects that might happen.

mniestroj replied:
@pabigot Just to note why there is no `IS_ENABLED()` in my commit, this is what gets generated when there is no `disconnect-sdo-sa0-pull-up` property specified:

```
#define DT_ST_STM32_SPI_FIFO_40013000_ST_LIS2DH_0_DISCONNECT_SDO_SA0_PULL_UP 0
#define DT_INST_0_ST_LIS2DH_DISCONNECT_SDO_SA0_PULL_UP DT_ST_STM32_SPI_FIFO_40013000_ST_LIS2DH_0_DISCONNECT_SDO_SA0_PULL_UP
```

pabigot replied:
Yes, is that a problem?

Using `IS_ENABLED(M)` is standard for Kconfig and devicetree macros that have 0 or 1 as their value.

It also seems to incidentally fix the build problem with frdm_k64f on test_build_sensors_i_z where there is no define for any LIS2DH devicetree entries.  I think you should put it back.

mniestroj replied:
You are right, I have put it back. For some reason I had a feeling that IS_ENABLED was always true (also in case of 0). I have retested this case once again and it works as you wrote (IS_ENABLED is false in case there is no dts flag specified).

avisconti commented on drivers/sensor/lis2dh/lis2dh.c:
```diff
@@ -321,6 +321,16 @@ int lis2dh_init(struct device *dev)
+		return status;
+	}
+#endif
+
```
Why always disabling the pullup? Shouldn't it be done only selectively depending from board to board?

avisconti replied:
There might be boards which actually requires the internal pullup to be set. So, I'm wondering if need to keep it configurable in Kconfig.

mniestroj replied:
My initial idea was to only disable it when the I2C address is `(addr & 0x1 == 0)`, i.e. 0x18 (which is the case on my HW). However this check would be valid only for I2C devices, not SPI. I think that SPI might have the same issue with power consumption, so it is good to disable that feature there as well. So lets consider those cases:
1) I2C addr is 0x18: we want to disable SDO/SA0 for sure, because we have pull-down on the board and a pull-up internally in the accelerometer,
2) I2C addr is 0x19: assuming that address is assigned during accelerometer power on, then we can either keep the pull-ups or disable them,
3) SPI: driving SDO pin will conflict a bit with pull-up on that line, which means that we might have increased power consumption on driving SDO line low.

Based on those I decided to always disable pull-up to simplify the code.

avisconti replied:
And why not leaving it configurable in Kconfig and disabled by default?
In this case you leave the possibility to enable it in case any board requires it.
It is just brain storming...

mniestroj replied:
Because it is not a software feature which we might want to enable or disable, such as triggering events on drdy. If we want to keep it configurable, then dts seems a better place, because there we actually describe how the hardware looks like. So we might introduce a `disable-sdo-sda0-pull-up` flag to control that behavior.

However from the 3 cases that I described above I am only a bit worried about case 2, i.e. whether the address is latched on applying power or it is rather "runtime" configuration. If it is latched, then we do not need to expose any configuration option to the user at all (either Kconfig or dts). If it is NOT latched, then we can add a condition `if (I2C && (addr & 0x1))`.

The reason I don't want to keep it configurable is simplicity for the end user. It is much easier to just enable the driver, configure in dts and expect it to perform well. If someone will need to examine which IC consumes too much power when the board has 10 sensors, then lot of configuration options and dts flags would need to be investigated before finding the one we want to introduce in this PR. And the last thing... I don't see any value of keeping it configurable, as there are limited use cases we can cover without exposing such option to user.

pabigot reviewed (changes requested):
I'm fine with the commit that unconditionally builds the helper function.

As for the pull-up control we should be avoiding Kconfig as devicetree is the "way forward".  For my suggested approach see below in context.

It would be nice to move the existing `irq-gpios` and proposed `disconnect-sdo-sda0-pull-up` into a new `st,lis2dh-common.yaml` file that is included by both the SPI and I2C variants.

mniestroj committed:
dts: bindings: lis2dh: create a common st,lis2dh-common.yaml

Create a common binding file that will be included by all bindings
handled by lis2dh.c driver. For now this includes optional irq-gpios
property.

Use introduced st,lis2dh-common.yaml in st,lsm303dlhc-accel.yaml in
order to support defining irq-gpios.

Also improve description of st,lis2dh-i2c.yaml to better match what can
be found in st,lis2dh-spi.yaml.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

mniestroj commented:
Based on the review comments I have introduced new device-tree property to disconnect SDO/SA0 pull-up when specified to be fully controlable by user. Hence previous logic based on compatible string has been dropped. Also a commit `dts: bindings: lis2dh: create a common st,lis2dh-common.yaml` has been cherry-picked from #20004 .

pabigot reviewed (changes requested):
Using IS_ENABLED(M) is standard for Kconfig and devicetree macros that have 0 or 1 as their value. It also evaluates to syntactically acceptable content when the macro definition is absent, as with frdm_k64f on test_build_sensors_i_z where there is no define for any LIS2DH devicetree entries. I think you should put it back which should fix the build problems.

avisconti reviewed (approved):
LGTM
